core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -157,21 +153,28 @@ class KafkaController(val config: KafkaConfig, zkUtils: ZkUtils, val brokerState
   val controllerContext = new ControllerContext(zkUtils)
   val partitionStateMachine = new PartitionStateMachine(this)
   val replicaStateMachine = new ReplicaStateMachine(this)
-  private val controllerElector = new ZookeeperLeaderElector(controllerContext, ZkUtils.ControllerPath, onControllerFailover,
-    onControllerResignation, config.brokerId, time)
   // have a separate scheduler for the controller to be able to start and stop independently of the
   // kafka server
-  private val autoRebalanceScheduler = new KafkaScheduler(1)
-  var deleteTopicManager: TopicDeletionManager = null
+  private val kafkaScheduler = new KafkaScheduler(1)
+  var topicDeletionManager: TopicDeletionManager = null
   val offlinePartitionSelector = new OfflinePartitionLeaderSelector(controllerContext, config)
   private val reassignedPartitionLeaderSelector = new ReassignedPartitionLeaderSelector(controllerContext)
   private val preferredReplicaPartitionLeaderSelector = new PreferredReplicaPartitionLeaderSelector(controllerContext)
   private val controlledShutdownPartitionLeaderSelector = new ControlledShutdownLeaderSelector(controllerContext)
   private val brokerRequestBatch = new ControllerBrokerRequestBatch(this)
 
-  private val partitionReassignedListener = new PartitionsReassignedListener(this)
-  private val preferredReplicaElectionListener = new PreferredReplicaElectionListener(this)
-  private val isrChangeNotificationListener = new IsrChangeNotificationListener(this)
+  private val controllerEventQueue = new LinkedBlockingQueue[ControllerEvent]
+  private val controllerEventThread = new ControllerEventThread("controller-event-thread")
+  private val brokerChangeListener = new BrokerChangeListener(this, controllerEventQueue)
+  private val topicChangeListener = new TopicChangeListener(this, controllerEventQueue)
+  private val topicDeletionListener = new TopicDeletionListener(this, controllerEventQueue)
+  private val partitionModificationsListeners: mutable.Map[String, PartitionModificationsListener] = mutable.Map.empty
+  private val partitionReassignmentListener = new PartitionReassignmentListener(this, controllerEventQueue)
+  private val preferredReplicaElectionListener = new PreferredReplicaElectionListener(this, controllerEventQueue)
+  private val isrChangeNotificationListener = new IsrChangeNotificationListener(this, controllerEventQueue)
```

Comment:
Sharing this logic at the listener layer would complicate the concurrency semantics since figuring out what has changed would require looking at controller local state while the listener is being executed from the ZkEventThread. This is what this PR is trying to avoid.

Even if we push the shared logic to the ControllerEvent layer, it doesn't seem like we'd actually simplify things since each listener reads and observes what's changed differently.